GCC 11.1 compilation issues

Hello,

I recently updated to GCC 11.1 and with this deal.ii compilation fails - basically because the implicitly created Tensor<> default move assignment operator gets implicitly deleted:


```
dealii::Tensor<1, 1, Sacado::Rad::ADvar<double> >& dealii::Tensor<1, 1, Sacado::Rad::ADvar<double> >::operator=(dealii::Tensor<1, 1, Sacado::Rad::ADvar<double> >&&)' is implicitly deleted because the default definition would be ill-formed:

[...]

include/deal.II/base/tensor.h:92:7: error: cannot bind non-const lvalue reference of type 'Sacado::Rad::ADvar<double>&' to an rvalue of type 'Sacado::Rad::ADvar<double>'
```

I have attached a longer extract of my build log here in this post.
This was not a problem even with GCC 10.3 - and at least as I see it the default constructor should not be implicitly deleted, but I suspect that we get this problem because of the overloaded assignment operators (which are already existent).

I could handle the issue by implementing some default construct / move operators by hand. With these changes deal.ii compiles and runs fine on GCC 11.1. You can see the changes in [this branch of my fork](https://github.com/mathsen/dealii/commit/8577359e44ae8cf02eea24f54358161c612f9d0f).

Some remarks: the for loop in the copy constructors is necessary because std::copy is not constexpr.
I deleted the ICC workaround because it isn't necessary anymore if we have to use the same workaround for GCC, too and implement them manually.

Maybe anybody has another idea how to cope with this issue?
Otherwise maybe the changes should be merged?

Greetings

Mathias


[log.txt](https://github.com/dealii/dealii/files/6503666/log.txt)

--- a/include/deal.II/base/tensor.h
+++ b/include/deal.II/base/tensor.h
@@ -923,7 +923,7 @@ constexpr DEAL_II_ALWAYS_INLINE DEAL_II_CUDA_HOST_DEV
 
 
 
-#if __GNUC__ >= 11 || defined __INTEL_COMPILER
+#  if __GNUC__ >= 11 || defined __INTEL_COMPILER
 template <int dim, typename Number>
 constexpr DEAL_II_ALWAYS_INLINE DEAL_II_CUDA_HOST_DEV
                                 Tensor<0, dim, Number>::Tensor(const Tensor<0, dim, Number> &other)
@@ -1282,7 +1282,7 @@ constexpr DEAL_II_ALWAYS_INLINE Tensor<rank_, dim, Number>::
 }
 
 
-#if __GNUC__ >= 11 || defined __INTEL_COMPILER
+#  if __GNUC__ >= 11 || defined __INTEL_COMPILER
 template <int rank_, int dim, typename Number>
 constexpr DEAL_II_ALWAYS_INLINE
 Tensor<rank_, dim, Number>::Tensor(const Tensor<rank_, dim, Number> &other)

--- a/include/deal.II/base/tensor.h
+++ b/include/deal.II/base/tensor.h
@@ -157,7 +157,6 @@ public:
   constexpr DEAL_II_CUDA_HOST_DEV
   Tensor(const Tensor<0, dim, OtherNumber> &initializer);
 
-#if __GNUC__ >= 11 || defined __INTEL_COMPILER
   /**
    * Constructor, where the data is copied from a C-style array.
    *
@@ -167,6 +166,7 @@ public:
   constexpr DEAL_II_CUDA_HOST_DEV
   Tensor(const OtherNumber &initializer);
 
+#if __GNUC__ >= 11 || defined __INTEL_COMPILER
   /**
    * Copy constructor
    */
@@ -923,7 +923,7 @@ constexpr DEAL_II_ALWAYS_INLINE DEAL_II_CUDA_HOST_DEV
 
 
 
-#  if __GNUC__ >= 11 || defined __INTEL_COMPILER
+#if __GNUC__ >= 11 || defined __INTEL_COMPILER
 template <int dim, typename Number>
 constexpr DEAL_II_ALWAYS_INLINE DEAL_II_CUDA_HOST_DEV
                                 Tensor<0, dim, Number>::Tensor(const Tensor<0, dim, Number> &other)
@@ -1282,7 +1282,7 @@ constexpr DEAL_II_ALWAYS_INLINE Tensor<rank_, dim, Number>::
 }
 
 
-#  if __GNUC__ >= 11 || defined __INTEL_COMPILER
+#if __GNUC__ >= 11 || defined __INTEL_COMPILER
 template <int rank_, int dim, typename Number>
 constexpr DEAL_II_ALWAYS_INLINE
 Tensor<rank_, dim, Number>::Tensor(const Tensor<rank_, dim, Number> &other)

--- a/include/deal.II/base/tensor.h
+++ b/include/deal.II/base/tensor.h
@@ -157,6 +157,7 @@ public:
   constexpr DEAL_II_CUDA_HOST_DEV
   Tensor(const Tensor<0, dim, OtherNumber> &initializer);
 
+#if __GNUC__ >= 11 || defined __INTEL_COMPILER
   /**
    * Constructor, where the data is copied from a C-style array.
    *
@@ -189,6 +190,7 @@ public:
    */
   constexpr DEAL_II_CUDA_HOST_DEV Tensor<0, dim, Number> &
                                   operator=(Tensor<0, dim, Number> &&other) noexcept;
+#endif
 
   /**
    * Return a pointer to the first element of the underlying storage.
@@ -558,6 +560,7 @@ public:
   constexpr
   operator Tensor<1, dim, Tensor<rank_ - 1, dim, OtherNumber>>() const;
 
+#if __GNUC__ >= 11 || defined __INTEL_COMPILER
   /**
    * Copy constructor
    */
@@ -579,6 +582,7 @@ public:
    */
   constexpr Tensor<rank_, dim, Number> &
   operator=(Tensor<rank_, dim, Number> &&) noexcept;
+#endif
 
   /**
    * Read-Write access operator.
@@ -919,6 +923,7 @@ constexpr DEAL_II_ALWAYS_INLINE DEAL_II_CUDA_HOST_DEV
 
 
 
+#  if __GNUC__ >= 11 || defined __INTEL_COMPILER
 template <int dim, typename Number>
 constexpr DEAL_II_ALWAYS_INLINE DEAL_II_CUDA_HOST_DEV
                                 Tensor<0, dim, Number>::Tensor(const Tensor<0, dim, Number> &other)
@@ -952,6 +957,7 @@ constexpr DEAL_II_ALWAYS_INLINE DEAL_II_CUDA_HOST_DEV Tensor<0, dim, Number> &
   value = std::move(other.value);
   return *this;
 }
+#  endif
 
 
 
@@ -1276,6 +1282,7 @@ constexpr DEAL_II_ALWAYS_INLINE Tensor<rank_, dim, Number>::
 }
 
 
+#  if __GNUC__ >= 11 || defined __INTEL_COMPILER
 template <int rank_, int dim, typename Number>
 constexpr DEAL_II_ALWAYS_INLINE
 Tensor<rank_, dim, Number>::Tensor(const Tensor<rank_, dim, Number> &other)
@@ -1313,6 +1320,7 @@ constexpr DEAL_II_ALWAYS_INLINE Tensor<rank_, dim, Number> &
     values[i] = other.values[i];
   return *this;
 }
+#  endif
 
 
 namespace internal

--- a/include/deal.II/base/tensor.h
+++ b/include/deal.II/base/tensor.h
@@ -167,6 +167,30 @@ public:
   Tensor(const OtherNumber &initializer);
 
   /**
+   * Copy constructor
+   */
+  constexpr DEAL_II_CUDA_HOST_DEV
+  Tensor(const Tensor<0, dim, Number> &other);
+
+  /**
+   * Copy assignment operator
+   */
+  constexpr DEAL_II_CUDA_HOST_DEV Tensor<0, dim, Number> &
+                                  operator=(const Tensor<0, dim, Number> &other);
+
+  /**
+   * Move constructor
+   */
+  constexpr DEAL_II_CUDA_HOST_DEV
+    Tensor(Tensor<0, dim, Number> &&other) noexcept;
+
+  /**
+   * Move assignment operator
+   */
+  constexpr DEAL_II_CUDA_HOST_DEV Tensor<0, dim, Number> &
+                                  operator=(Tensor<0, dim, Number> &&other) noexcept;
+
+  /**
    * Return a pointer to the first element of the underlying storage.
    */
   Number *
@@ -222,19 +246,6 @@ public:
   template <typename OtherNumber>
   constexpr DEAL_II_CUDA_HOST_DEV Tensor &
                                   operator=(const Tensor<0, dim, OtherNumber> &rhs);
-
-#ifdef __INTEL_COMPILER
-  /**
-   * Assignment from tensors with same underlying scalar type.
-   * This is needed for ICC15 because it can't generate a suitable
-   * copy constructor for Sacado::Rad::ADvar types automatically.
-   * See https://github.com/dealii/dealii/pull/5865.
-   *
-   * @note This function can also be used in CUDA device code.
-   */
-  constexpr DEAL_II_CUDA_HOST_DEV Tensor &
-                                  operator=(const Tensor<0, dim, Number> &rhs);
-#endif
 
   /**
    * This operator assigns a scalar to a tensor. This obviously requires
@@ -546,6 +557,28 @@ public:
   template <typename OtherNumber>
   constexpr
   operator Tensor<1, dim, Tensor<rank_ - 1, dim, OtherNumber>>() const;
+
+  /**
+   * Copy constructor
+   */
+  constexpr Tensor(const Tensor<rank_, dim, Number> &);
+
+  /**
+   * Copy assignment operator
+   */
+  constexpr Tensor<rank_, dim, Number> &
+  operator=(const Tensor<rank_, dim, Number> &);
+
+  /**
+   * Move constructor
+   */
+  constexpr Tensor(Tensor<rank_, dim, Number> &&) noexcept;
+
+  /**
+   * Move assignment operator
+   */
+  constexpr Tensor<rank_, dim, Number> &
+  operator=(Tensor<rank_, dim, Number> &&) noexcept;
 
   /**
    * Read-Write access operator.
@@ -887,6 +920,42 @@ constexpr DEAL_II_ALWAYS_INLINE DEAL_II_CUDA_HOST_DEV
 
 
 template <int dim, typename Number>
+constexpr DEAL_II_ALWAYS_INLINE DEAL_II_CUDA_HOST_DEV
+                                Tensor<0, dim, Number>::Tensor(const Tensor<0, dim, Number> &other)
+  : value{other.value}
+{}
+
+
+
+template <int dim, typename Number>
+constexpr DEAL_II_ALWAYS_INLINE DEAL_II_CUDA_HOST_DEV Tensor<0, dim, Number> &
+Tensor<0, dim, Number>::operator=(const Tensor<0, dim, Number> &other)
+{
+  value = other.value;
+  return *this;
+}
+
+
+
+template <int dim, typename Number>
+constexpr DEAL_II_ALWAYS_INLINE DEAL_II_CUDA_HOST_DEV
+                                Tensor<0, dim, Number>::Tensor(Tensor<0, dim, Number> &&other) noexcept
+  : value{std::move(other.value)}
+{}
+
+
+
+template <int dim, typename Number>
+constexpr DEAL_II_ALWAYS_INLINE DEAL_II_CUDA_HOST_DEV Tensor<0, dim, Number> &
+  Tensor<0, dim, Number>::operator=(Tensor<0, dim, Number> &&other) noexcept
+{
+  value = std::move(other.value);
+  return *this;
+}
+
+
+
+template <int dim, typename Number>
 inline Number *
 Tensor<0, dim, Number>::begin_raw()
 {
@@ -957,18 +1026,6 @@ constexpr inline DEAL_II_ALWAYS_INLINE
   value = internal::NumberType<Number>::value(p);
   return *this;
 }
-
-
-#  ifdef __INTEL_COMPILER
-template <int dim, typename Number>
-constexpr inline DEAL_II_ALWAYS_INLINE
-  DEAL_II_CUDA_HOST_DEV Tensor<0, dim, Number> &
-  Tensor<0, dim, Number>::operator=(const Tensor<0, dim, Number> &p)
-{
-  value = p.value;
-  return *this;
-}
-#  endif
 
 
 template <int dim, typename Number>
@@ -1218,6 +1275,44 @@ constexpr DEAL_II_ALWAYS_INLINE Tensor<rank_, dim, Number>::
   return Tensor<1, dim, Tensor<rank_ - 1, dim, Number>>(values);
 }
 
+
+template <int rank_, int dim, typename Number>
+constexpr DEAL_II_ALWAYS_INLINE
+Tensor<rank_, dim, Number>::Tensor(const Tensor<rank_, dim, Number> &other)
+{
+  for (unsigned int i = 0; i < dim; ++i)
+    values[i] = other.values[i];
+}
+
+
+template <int rank_, int dim, typename Number>
+constexpr DEAL_II_ALWAYS_INLINE Tensor<rank_, dim, Number> &
+Tensor<rank_, dim, Number>::operator=(const Tensor<rank_, dim, Number> &other)
+{
+  for (unsigned int i = 0; i < dim; ++i)
+    values[i] = other.values[i];
+  return *this;
+}
+
+
+template <int rank_, int dim, typename Number>
+constexpr DEAL_II_ALWAYS_INLINE
+Tensor<rank_, dim, Number>::Tensor(Tensor<rank_, dim, Number> &&other) noexcept
+{
+  for (unsigned int i = 0; i < dim; ++i)
+    values[i] = other.values[i];
+}
+
+
+template <int rank_, int dim, typename Number>
+constexpr DEAL_II_ALWAYS_INLINE Tensor<rank_, dim, Number> &
+                                Tensor<rank_, dim, Number>::
+                                operator=(Tensor<rank_, dim, Number> &&other) noexcept
+{
+  for (unsigned int i = 0; i < dim; ++i)
+    values[i] = other.values[i];
+  return *this;
+}
 
 
 namespace internal

